File: Complementar - Estruturas de dados (Didático)/Algoritmos15/Alg15.php
<?php
/*
♦ Algoritmo 15: Estrutura de Dados Linear Dinâmica (não Estática) de Pilha: Stack (Pilha).
*/

class PilhaDinamica {
    private $pilha;

    public function __construct() {
        $this->pilha = [];
    }

    public function isEmpty() {
        return empty($this->pilha);
    }

    public function push($elemento) {
        array_unshift($this->pilha, $elemento);
    }

    public function pop() {
        if ($this->isEmpty()) {
            throw new Exception("A pilha está vazia");
        }
        array_shift($this->pilha);
    }

    public function size() {
        return count($this->pilha);
    }

    public function front() {
        if ($this->isEmpty()) {
            throw new Exception("A pilha está vazia");
        }
        return $this->pilha[0];
    }

    public function showStack() {
        echo implode(' ', $this->pilha) . PHP_EOL;
    }
}

$pilha = new PilhaDinamica();

$pilha->push(1);
$pilha->push(2);
$pilha->push(3);

$pilha->showStack();

echo "Tamanho da pilha: " . $pilha->size() . PHP_EOL;
echo "Topo da pilha: " . $pilha->front() . PHP_EOL;

$pilha->pop();
$pilha->showStack();

?>

File: Complementar - Estruturas de dados (Didático)/Algoritmos15/Alg15.py
'''
♦ Algoritmo 15: Estrutura de Dados Linear Dinâmica (não Estática) de Pilha: Stack (Pilha).
'''

class PilhaDinamica:
    def __init__(self):
        self.pilha = []

    def is_empty(self):
        return len(self.pilha) == 0

    def push(self, elemento):
        self.pilha.insert(0, elemento)

    def pop(self):
        if self.is_empty():
            raise Exception("A pilha está vazia")
        self.pilha.pop(0)

    def size(self):
        return len(self.pilha)

    def front(self):
        if self.is_empty():
            raise Exception("A pilha está vazia")
        return self.pilha[0]

    def show_stack(self):
        print(" ".join(map(str, self.pilha)))

pilha = PilhaDinamica()

pilha.push(1)
pilha.push(2)
pilha.push(3)

pilha.show_stack()

print("Tamanho da pilha:", pilha.size())
print("Topo da pilha:", pilha.front())

pilha.pop()
pilha.show_stack()
